File: packages/react-native-audio-api/ios/nodes/AudioNode/AudioNode.m
#import "AudioNode.h"
#import "AudioContext.h"

@implementation AudioNode

- (instancetype)initWithContext:(AudioContext *)context
{
  if (self = [super init]) {
    _connectedNodes = [NSMutableArray array];
    _context = context;
    _channelCount = 2;
    _channelCountMode = ChannelCountModeMax;
    _channelInterpretation = ChannelInterpretationSpeakers;
  }

  return self;
}

- (void)cleanup
{
  _context = nil;
  [_connectedNodes removeAllObjects];
  _connectedNodes = nil;
}

- (void)process:(AVAudioFrameCount)frameCount bufferList:(AudioBufferList *)bufferList;
{
  for (AudioNode *node in _connectedNodes) {
    [node process:frameCount bufferList:bufferList];
  }
}

- (void)connect:(AudioNode *)node
{
  if (_numberOfOutputs > 0) {
    [_connectedNodes addObject:node];
  }
}

- (void)disconnect:(AudioNode *)node
{
  NSUInteger index = [_connectedNodes indexOfObject:node];
  if (index != NSNotFound) {
    [_connectedNodes removeObjectAtIndex:index];
  }
}

- (int)getNumberOfInputs
{
  return _numberOfInputs;
}

- (int)getNumberOfOutputs
{
  return _numberOfOutputs;
}

- (int)getChannelCount
{
  return _channelCount;
}

- (NSString *)getChannelCountMode
{
  return [ChannelCountMode stringFromChannelCountMode:self.channelCountMode];
}

- (NSString *)getChannelInterpretation
{
  return [ChannelInterpretation stringFromChannelInterpretation:self.channelInterpretation];
}

@end

File: packages/react-native-audio-api/android/src/main/cpp/AudioNode/AudioNode.cpp
#include "AudioNode.h"
#include "AudioContext.h"

namespace audioapi {

AudioNode::AudioNode(AudioContext *context) : context_(context) {}

AudioNode::~AudioNode() {
  cleanup();
}

int AudioNode::getNumberOfInputs() const {
  return numberOfInputs_;
}

int AudioNode::getNumberOfOutputs() const {
  return numberOfOutputs_;
}

int AudioNode::getChannelCount() const {
  return channelCount_;
}

std::string AudioNode::getChannelCountMode() const {
  return AudioNode::toString(channelCountMode_);
}

std::string AudioNode::getChannelInterpretation() const {
  return AudioNode::toString(channelInterpretation_);
}

void AudioNode::connect(const std::shared_ptr<AudioNode> &node) {
  if (numberOfOutputs_ > outputNodes_.size() &&
      node->getNumberOfInputs() > node->inputNodes_.size()) {
    outputNodes_.push_back(node);
    node->inputNodes_.push_back(shared_from_this());
  }
}

void AudioNode::disconnect(const std::shared_ptr<AudioNode> &node) {
  outputNodes_.erase(
      std::remove(outputNodes_.begin(), outputNodes_.end(), node),
      outputNodes_.end());
  if (auto sharedThis = shared_from_this()) {
    node->inputNodes_.erase(
        std::remove(
            node->inputNodes_.begin(), node->inputNodes_.end(), sharedThis),
        node->inputNodes_.end());
  }
}

void AudioNode::cleanup() {
  outputNodes_.clear();
  inputNodes_.clear();
}

bool AudioNode::processAudio(float *audioData, int32_t numFrames) {
  bool isPlaying = false;
  for (auto &node : inputNodes_) {
    if (node->processAudio(audioData, numFrames)) {
      isPlaying = true;
    }
  }

  return isPlaying;
}

} // namespace audioapi
